feature: blog creation and editing implemented

amaazahmad opened the pull request.

amaazahmad committed:
TYPE: feature

SUBJECT: Basic components for Form + Routing for Login/signup

DESCRIPTION: InputBar and Form components have been created to be used in both login and signup forms. Default routing set to the login page using <Navigate/>. Routing between Signup and Login pages settled.

amaazahmad committed:
TYPE: Refactor

SUBJECT: Deleted unused files. Changed the method of accessing form elements in userForm.

DESCRIPTION: Deleted a few default files which were not being used e.g. webVitals, etc. Used the FormEvent instead of SyntheticEvent in userForm component and changed the method of accessing form properties.

amaazahmad committed:
TYPE: feature

SUBJECT: Firebase user authentication applied + HTML form replaced with react-hook-form

DESCRIPTION: Firebase authentication with email/password applied. Still not able to store the name of the user, though. The retrieved user object is yet to be stored in Context and LocalStorage. The HTML form element in userForm now uses react-hook-form.

amaazahmad committed:
TYPE: fix

SUBJECT: Password and Confirm Password validation added in react hook form.

amaazahmad committed:
TYPE: feature-improvement

SUBJECT: Signout functionality implemented. Routing based on user logged-in status.

DESCRIPTION: Signout functionality completely implemented. Didn't use local storage as Firebase keeps the track of logged-in user with auth object. Routing implemented: default page set to login if the user is not logged-in, default page is home if the user is logged-in.

amaazahmad committed:
TYPE: styles

SUBJECT: Login and Signup pages styling for mobile display

DESCRIPTION: Added styling for the login and signup pages by creating CSS files for now, with a fixed screen size in px. Responsiveness and Tailwind yet to be done.

amaazahmad committed:
TYPE: styling fix

SUBJECT: Placeholder vs text color fixed for input fields in login/signup forms.

amaazahmad committed:
TYPE: styles

SUBJECT: Responsive settled and login/signup screens made for larger screens

DESCRIPTION: Added a side-bar for login/signup form for larger screens. Container divs added in login and signup pages for CSS implementation.

YET TO BE DONE:
- Use Tailwind instead of CSS files
- Adjust the heading size in the left sidebar of login/signup page for medium sized screens. It stacks on different lines for med screens.

amaazahmad committed:
TYPE: Styling-fixes

SUBJECT: Styling fixes done

DESCRIPTION: Following styling fixes done in login/signup pages:
- Left aligned welcome heading and subheading for desktop screen
- Input fields max-width adjusted
- Form's padding adjusted
- Font-size readjusted for headings and input fields
- LeftImageContainer size increased from 33% to 37%

amaazahmad committed:
TYPE: Styling fixes

SUBJECT: Tiny styling improvements for responsiveness

amaazahmad committed:
TYPE: styling-fix

SUBJECT: Styling improved after first demo

DESCRIPTION:
- Showed user not found error instead of incorrect password error. FIXED
- Increased padding with left side of form: Changed to 8rem from 4rem after 1024px screen width
- Error font size slightly increased.

amaazahmad committed:
style: Tailwind CSS implemented

Description: All CSS files have been removed and tailwind has been implemented. An unused InputBar component has been deleted.

amaazahmad committed:
refactor: fixed issues raised in the first PR

Description:
- simple functions changed to arrow functions
- comments and log statements removed
- props destructred in the function definition
- imports sorted
- firebase config shifted to .env.local

amaazahmad committed:
fix: Fixed the flashing of login screen if a user is logged and refreshes the home page

description: added a onAuthStateChanged listener in the useEffect hook of the App.tsx. This reduces the delay caused by setting context n then reading the currentUser from there.

amaazahmad committed:
fix+refactor: Made the form component generic and Login screen flashing fixed

Description:
- The login screen used to flash for a second if a logged-in user refreshes the application. That was fixed by using the onAuthStateChanged listener directly in the useEffect hook.
- The form component was made generic and then used in both Login and Sign up pages.

amaazahmad committed:
refactor: minor improvements in error messages and data type for react-hook-form

description:
- Used the FieldValues data type in place of any type on the submitHandler function in react-hook-form
- Removed the validation rule for confirm password and password from the confirmPassword field. This is now done in onSubmitHandler
- Moved the errorMessage for firebase errors outside the form, in the login and signup components independently.

amaazahmad committed:
feat: Blog dashboard on home page implemented

description:
- New blogEntry, blogList and sidebar components created
- styling is done in CSS files for now. To be shifted to tailwind
- separate utils file made for utils relevant to firebase firebaseDB
- signout function yet to be implemented for mobile and tab screens

amaazahmad committed:
feeature: MyBlogs dashboard implemented

description: Clicking on the user icon in sidebar leads to the blogs of the signed-in user. Optional parameter passed to the firebase util function getBlogs

amaazahmad committed:
minor fix: renamed the function in MyBlogs

amaazahmad committed:
styling: User's name is now displayed on the bottom bar in tab view and initial letter in the user icon.

amaazahmad committed:
styling: All CSS files deleted and styling moved to tailwind

amaazahmad committed:
styling fix: minor fix in height of Outlet in Sidebar

amaazahmad committed:
refactor: spacing and indentation fixed with default formatter of VS code

amaazahmad committed:
feature: blog detail view implemented

description:
- Data retrieved from firebase changed. Unique key generated by firebase also included now.
- Routing adjusted. /blog/:id is the link to each blog's detail view
- BlogView styling is still in CSS files. Will be converted to Tailwind once the position of edit and delete buttons are final

amaazahmad committed:
feature: dark mode implemented + partial search functionality implemented

description:
- dark mode implemented
- Context made for storing the theme mode. Theme is also stored in localStorage to be obtained on refresh.
- search bar functionality is fine. Two minor issues:
     1. Placement n styling?
     2. It does not close when clicked on the borders of the white window
- some commented code left on the Home page intentionally. It is a part of the search bar functionality.

amaazahmad committed:
styling: signout button placed for tab n mobile screens

amaazahmad committed:
styling: search bar visibility and position adjusted

amaazahmad committed:
fix: minor bug fix for blogView display for laptop screens
refactor: blog.d.ts file created in types folder to be exported wherever BlogData is needed

amaazahmad committed:
styling: hover cursor adjusted for sidebar search icon

amaazahmad committed:
refactor: functionality to get theme from local storage moved to ThemeProvider's useEffect from App.tsx useEffect

amaazahmad committed:
fix: Routing set to login if a user tries to view blogs without logging in. Routing set to home page if a logged-in user tries to access login/signup pages

amaazahmad committed:
styling: Minor changes + Styling implemented in tailwind

description:
- Reduced the white shadow of sidebar in dark mode (suggested by client)
- Reduced the paddings of blog view in mobile view (suggested by client)
- Implemented styling in Tailwind for the BlogView component and Home page.
- Delete and Edit icons placed in the blog list of My Blogs page

amaazahmad committed:
fix: MyBlogs page adjusted

amaazahmad committed:
feature/style: cross icon added in the search bar (suggested by the client)

description:
- The cross icon clears the search field if there is any string present
- If the search field is empty, the cross icon removes the search field from the view

amaazahmad committed:
feature: delete confirmation pop-up delete

description:
- Added a pop-up modal when the delete icon is clicked
- delete blog function added in firebaseDB utils

amaazahmad committed:
styling: deleted css file for deleteConfirmation component and implemented tailwind

amaazahmad committed:
styling: dark mode implemented for the deleteConfirmation popup

amaazahmad committed:
feature: Pop-ups set on click of edit icon n create icon

amaazahmad committed:
style: Added styling for the blog creation/edit popup

amaazahmad committed:
feature: Deletion and Creation of a new blog implemented

amaazahmad committed:
refactor: warnings removed from firebase utils

amaazahmad committed:
feature: blog editing feature implemented

amaazahmad committed:
styles: new line character displayed using the white-space:pre-wrap CSS property

amaazahmad committed:
styles: several improvements in styling

description:
- edit/create blog pop-up width reduced to 55%
- For edit blog pop-up, button text changed to "Save" and button right aligned
- For create blog pop-up, button text changed to "Create" and button right aligned
- In deleteConfirmation pop-up, width of Cancel and Delete button slightly increased
- Left and right padding in blog list decreased for mobile view
- Internal padding for the edit/create pop-up decreased for mobile screens
- delete pop-up width increased for mobile screens
- Moved the Delete button above the Cancel button in deleteConfirmation pop-up for mobile screens

amaazahmad committed:
feature: the create/edit pop-ups do not close on clicking the overlay. A close button has been added to the top-tight corner to avoid accidental clicks

amaazahmad committed:
style: Loaders added in the add/save/delete buttons

amaazahmad committed:
style: minor styling fix for loader in buttons

amaazahmad committed:
style: loader implemented in home page and my blogs page

amaazahmad committed:
refactor: warnings fixed

amaazahmad committed:
navigation added to the blog view after edit is done

amaazahmad committed:
loader added if a blog view is directly opened

amaazahmad committed:
styling: multiple fixes, as suggested in the final demo

description:
- Loader for the blog view placed in the center of the screen
- If no blog matches the search criteria, a message is now displayed
- Size of the warning icon slightly reduced in the deleteConfirmation popup
- Size of the secondary text slightly increased in the deleteConfirmation popup
- Buttons placed in the same row in deleteConfirmation pop-up and text size matches that of the secondary text in the same popup
- Cancel button placed on the left of the Save/Add button in create/edit popup. Cancel button, with full width, is below the save button for mobile screens.
- Edit pop-up and Create-popup now follows the same styling
- Line heights for blog view page adjusted to be more closer to the designs
- Attempted to make padding consistent for the entire site
- Warning modal for signout implemented. It is same as the deleteConfirmation modal.

amaazahmad committed:
refactor: console logs removed and imports ordered

amaazahmad committed:
Close icon also placed on the top-right corner of the deleteConfirmation popup modal

amaazahmad committed:
fixed issues raised in the second review

amaazahmad committed:
style: signout popup modal styling fixed

amaazahmad committed:
refactor: minor changes

description:
- minor changes in destructuring of props
- addedd efault styling for the form component
- ordered imports
- self-closing tags added where possible

maryamrasool commented on src/App.tsx:
```diff
@@ -1,26 +1,51 @@
+					<Route index={true} element={<HomePage />} />
+					<Route path="my-blogs" element={<MyBlogs />} />
+				</Route>
+				<Route path="/login" element={<LoginPage />}></Route>
```
When there is nothing to add in the tag, it is better to make it self closing.
```suggestion
				<Route path="/login" element={<LoginPage />} />
```

amaazahmad replied:
My bad, sorry. 

Fixed.

maryamrasool commented on src/components/blogList/blogList.tsx:
```diff
@@ -0,0 +1,40 @@
+}
+
+const BlogList = ({ blogs, isMyBlogs }: BlogListProps) => {
+	const [pageState, setPageState] = useState<string>("")
```
This state variable is not required. Check out the comment on the useEffect hook.

amaazahmad replied:
Fixed.

amaazahmad replied:
pageState removed. The message is displayed by the Home/My Blogs page now.

maryamrasool commented on src/components/blogList/blogList.tsx:
```diff
@@ -0,0 +1,40 @@
+	const { theme } = useContext(ThemeContext)
+
+	useEffect(() => {
+		if (blogs?.length === 0) setPageState("No blogs found.")
```
You are checking the length of the blogs array to see if any blogs exist. In case, there is no blog, you'll set the page state to `No Blogs Found` and render it.
This is an unnecessary step, utilising a state variable and a useEffect hook that will trigger multiple renders and slow down your application.
In the render function, you are again checking the length of the blogs array and showing this state variable. 
How about we not set the state variable in the hook and directly display the `No Blogs Found` message after we are checking the blogs array length in the render function?

amaazahmad replied:
It served another purpose initially, when there was no loader. 
My bad. 

I've removed it.

maryamrasool commented on src/components/blogList/blogList.tsx:
```diff
@@ -0,0 +1,40 @@
+				</div>
+			) : (
+				<div className="w-full flex justify-center items-center m-0">
+					<p className={`${theme ? "text-white" : "text-dark-gray-text-color"} pl-8 pt-8 text-xl font-normal font-lexend-deca box-border`}>{pageState}</p>
```
Remove the pageState state variable and the useEffect used to set that state variable and directly use `No Blogs Found` message here.

amaazahmad replied:
Done.

maryamrasool commented on src/components/blogListEntry/blogListEntry.tsx:
```diff
@@ -0,0 +1,113 @@
+import { useEffect, useState, useContext } from 'react';
```
Good job with organising your imports.

maryamrasool commented on src/components/blogListEntry/blogListEntry.tsx:
```diff
@@ -0,0 +1,113 @@
+					{dateToDisplay}
+				</p>
+				:
+				<></>
```
Instead of rendering an empty tag, it's better to write null here. This occurs more than once across the repository. Can you fix it everywhere?

amaazahmad replied:
Sure. 
I thought its ok to leave an empty fragment cuz fragments are not converted to HTML any way. 

I'll fix this right away.

maryamrasool commented on src/components/blogListEntry/blogListEntry.tsx:
```diff
@@ -0,0 +1,113 @@
+						open={editPopup}
+						onOpen={() => { setEditPopup(true) }}
+						onClose={() => { setEditPopup(false) }}
+						contentStyle={screenWidth >= 768 ? { padding: 0, border: 0, width: '75%', height: '70%' } : { padding: 0, border: 0, width: '75%', height: '70%' }}
```
This conditional operation returns the same value whether the condition is "true" or "false".

amaazahmad replied:
I'm so sorry for this silly mistake 😂
I've fixed it.

maryamrasool commented on src/components/deleteConfirmation/deleteConfirmation.component.tsx:
```diff
@@ -0,0 +1,51 @@
+const DeleteConfirmation = ({ blogID, setDeletePopup }: DeleteConfirmationProps) => {
+
+     const { theme } = useContext(ThemeContext);
+     const [loaderVisible, setLoaderVisible] = useState<boolean>(false);
```
With boolean variables, it's better to name your variables like this: is{variableName}, so in this case, it would be `isLoaderVisible`.

This goes for all the boolean variables. Can you check in the whole repository?

amaazahmad replied:
Done.

maryamrasool commented on src/components/editBlog/editBlog.component.tsx:
```diff
@@ -0,0 +1,92 @@
+                    message: 'Title can not be empty.'
+               },
+               defaultValue: blog?.title,
+               styles: `${theme ? "text-white bg-dark-gray-text-color border-green-text-color" : "bg-white text-gray-900 border-secondary-text-color focus:border-green-text-color"} box-border w-full not-italic  outline-none leading-5 text-base m-0 p-4 border-2 border-solid  font-lexend-deca  placeholder-secondary-text-color w-[110%]
```
The styles are mostly same for all the fields. As discussed in the call, move these styles to the form component and modify it to accept some custom styles. Custom styles will be passed in few cases where the input field is different.

amaazahmad replied:
The form component had default styles, if no style property was passed for any field. 
If a style field was passed, the entire default style was over-written.

Now, the default styles are concatenated with the provided styles (if any). This way, default styles persist unless over-written by a custom style property.

maryamrasool commented on src/components/editBlog/editBlog.component.tsx:
```diff
@@ -0,0 +1,92 @@
+     const onSubmitHandler = async ({ title, content }: FieldValues) => {
+          if (blog) {
+               //updating an existing blog
+               const response = await updateBlog(blog.key, title, content)
```
Can we move the if block code to a separate function and just call that function here? It will be simpler and the function of the name will tell that we are updating the blog. We would not have to add a comment here.
As a note for future, try to make smaller functions and make sure that one function has only one job. In our case, handleSubmit should decide whether we are updating an existing blog or adding a new blog and just call the respective function for that.

amaazahmad replied:
Done.

maryamrasool commented on src/components/form/form.component.tsx:
```diff
@@ -0,0 +1,130 @@
+			<form
+				className="flex flex-col w-4/5 justify-center items-center h-full sm:w-full "
+				onSubmit={handleSubmit(async (data) => {
+					await onSubmitHandler(data)
```
Async and Await are not required here. They are handled within the submitHandler function passed from the parent component.

amaazahmad replied:
Done.

maryamrasool commented on src/context/theme/theme.provider.tsx:
```diff
@@ -0,0 +1,24 @@
+export const ThemeProvider: React.FC<ThemeProviderProps> = ({
+     children,
+}: ThemeProviderProps) => {
+     const [theme, setTheme] = useState<boolean>(false)
```
The name for this state variable is not appropriate. You're checking if user has selected dark theme or not. So a better name here would be `isDarkTheme`

amaazahmad replied:
Done

maryamrasool commented on src/pages/blogView/blogView.component.tsx:
```diff
@@ -0,0 +1,83 @@
+     const location = useLocation()
+     const navigate = useNavigate();
+
+     const [title, setTitle] = useState<string>("")
```
Creating individual state variables for all the fields of the blog is not the right approach. Each time a state variable is set, rendering of the component is triggered. With this approach, you are triggering the render function unnecessarily. 
A better approach here would be to create one state variable called `blog` and use dot notation to access the property (title, content and etc) of the blog.

amaazahmad replied:
![Screenshot 2022-10-10 at 4 04 39 PM](https://user-images.githubusercontent.com/111420831/194852388-676cfdf6-3d46-4ef2-b6e2-cf5d75eed384.png)

I'll change it like you suggested.

However, I did check this with console logs.
I believe the component is rendered only once, with useEffect being called twice. I set all the states simultaneously, so react does a single re-render.

amaazahmad replied:
Change implemented.

maryamrasool commented on src/pages/blogView/blogView.component.tsx:
```diff
@@ -0,0 +1,83 @@
+     }
+
+     useEffect(() => {
+          const getBlogData = async () => {
```
It's better to declare functions outside useEffect and just call them within useEffect.
Also, the content of this function could be broken down to create at least two smaller functions. I suggest dividing the if else code into separate functions.

amaazahmad replied:
WIth a single state variable, each if/else code block is just 2-3 lines. I think it'd be good enough as it is. 

However, I'll make the changes if you still suggest to divide it into functions.

amaazahmad committed:
type: refactor

description:
- Naming convention made consistent for icons
- App.css removed and styling applied as tailwind classes to App component
- App.test.tsx removed
- App.tsx function changed to arrow function from normal function
- React.env.ts removed

amaazahmad committed:
feature/refactor: The entire page does not reload when a blog is deleted. The data is fetched again and only the list is updated.

amaazahmad committed:
minor fix in styling of blog list

amaazahmad committed:
All changed suggested in PR review implemented except styling of form

amaazahmad committed:
styling adjusted a bit for the form and custom styling for edit blog component

amaazahmad committed:
styling fix: minor fix in title field of edit/create popup

amaazahmad committed:
refactor: The 'No Blogs Found' message is now displayed by the Home and My Blogs page instead of the blogList component. The page state from the BlogList component has been removed.
